File: temporal/worker-go/activities/saga_activities.go
package activities

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"go.temporal.io/sdk/activity"
)

// ReserveInventory simulates reserving an item in inventory.
func ReserveInventory(ctx context.Context, orderID string, item string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("ReserveInventory", "orderId", orderID, "item", item)

	time.Sleep(800 * time.Millisecond)

	logger.Info(fmt.Sprintf("📦 Inventory reserved: %s for order %s", item, orderID))
	return fmt.Sprintf("reserved:%s:%s", item, orderID), nil
}

// ChargePayment simulates charging a payment.
func ChargePayment(ctx context.Context, orderID string, amount float64) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("ChargePayment", "orderId", orderID, "amount", amount)

	time.Sleep(1 * time.Second)

	logger.Info(fmt.Sprintf("💳 Payment charged: $%.2f for order %s", amount, orderID))
	return fmt.Sprintf("charged:%.2f:%s", amount, orderID), nil
}

// ShipOrder simulates shipping an order. Has a ~30% simulated failure rate
// to demonstrate saga compensation.
func ShipOrder(ctx context.Context, orderID string, item string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("ShipOrder", "orderId", orderID, "item", item)

	time.Sleep(1 * time.Second)

	// Simulate occasional shipping failure (~30% chance)
	// Uses workflow-deterministic randomness via activity info
	info := activity.GetInfo(ctx)
	r := rand.New(rand.NewSource(time.Now().UnixNano() + int64(info.Attempt)))
	if r.Float64() < 0.3 {
		logger.Warn("ShipOrder FAILED (simulated)", "orderId", orderID)
		return "", fmt.Errorf("shipping failed: warehouse unavailable for order %s", orderID)
	}

	logger.Info(fmt.Sprintf("🚚 Order shipped: %s (%s)", orderID, item))
	return fmt.Sprintf("shipped:%s:%s", item, orderID), nil
}

// CancelReservation compensates for ReserveInventory.
func CancelReservation(ctx context.Context, orderID string, item string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("CancelReservation (compensation)", "orderId", orderID, "item", item)

	time.Sleep(500 * time.Millisecond)

	logger.Info(fmt.Sprintf("↩️ Reservation cancelled: %s for order %s", item, orderID))
	return fmt.Sprintf("cancelled-reservation:%s:%s", item, orderID), nil
}

// RefundPayment compensates for ChargePayment.
func RefundPayment(ctx context.Context, orderID string, amount float64) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("RefundPayment (compensation)", "orderId", orderID, "amount", amount)

	time.Sleep(500 * time.Millisecond)

	logger.Info(fmt.Sprintf("💸 Payment refunded: $%.2f for order %s", amount, orderID))
	return fmt.Sprintf("refunded:%.2f:%s", amount, orderID), nil
}

File: temporal/worker-ts/src/activities/sagaActivities.ts
import { log } from '@temporalio/activity';

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// --- Forward steps ---

export async function reserveInventory(orderId: string, item: string): Promise<string> {
  log.info('reserveInventory', { orderId, item });
  await delay(800);
  log.info(`📦 Inventory reserved: ${item} for order ${orderId}`);
  return `reserved:${item}:${orderId}`;
}

export async function chargePayment(orderId: string, amount: number): Promise<string> {
  log.info('chargePayment', { orderId, amount });
  await delay(1000);
  log.info(`💳 Payment charged: $${amount.toFixed(2)} for order ${orderId}`);
  return `charged:${amount.toFixed(2)}:${orderId}`;
}

export async function arrangeShipping(orderId: string, item: string): Promise<string> {
  log.info('arrangeShipping', { orderId, item });
  await delay(1000);
  log.info(`🚚 Shipping arranged: ${item} for order ${orderId}`);
  return `shipping-arranged:${item}:${orderId}`;
}

/**
 * sendConfirmation — has a ~30% failure rate to trigger compensations.
 */
export async function sendConfirmation(
  orderId: string,
  item: string,
  amount: number
): Promise<string> {
  log.info('sendConfirmation', { orderId, item, amount });
  await delay(500);

  // Simulate occasional failure (~30% chance)
  if (Math.random() < 0.3) {
    log.warn('sendConfirmation FAILED (simulated)', { orderId });
    throw new Error(`Confirmation service unavailable for order ${orderId}`);
  }

  log.info(`✉️ Confirmation sent: order ${orderId}, ${item}, $${amount.toFixed(2)}`);
  return `confirmed:${orderId}`;
}

// --- Compensation steps ---

export async function cancelReservation(orderId: string, item: string): Promise<string> {
  log.info('cancelReservation (compensation)', { orderId, item });
  await delay(500);
  log.info(`↩️ Reservation cancelled: ${item} for order ${orderId}`);
  return `cancelled-reservation:${item}:${orderId}`;
}

export async function refundPayment(orderId: string, amount: number): Promise<string> {
  log.info('refundPayment (compensation)', { orderId, amount });
  await delay(500);
  log.info(`💸 Payment refunded: $${amount.toFixed(2)} for order ${orderId}`);
  return `refunded:${amount.toFixed(2)}:${orderId}`;
}

/**
 * cancelShipping — TS-specific: this compensation itself has a ~20% failure chance.
 * Demonstrates that compensations can fail too, and the saga handles it gracefully.
 */
export async function cancelShipping(orderId: string, item: string): Promise<string> {
  log.info('cancelShipping (compensation)', { orderId, item });
  await delay(500);

  // Simulate occasional compensation failure
  if (Math.random() < 0.2) {
    log.error('cancelShipping FAILED (simulated)', { orderId });
    throw new Error(`Shipping cancellation failed for order ${orderId} — manual intervention needed`);
  }

  log.info(`🚫 Shipping cancelled: ${item} for order ${orderId}`);
  return `cancelled-shipping:${item}:${orderId}`;
}
